src/operator/optimizer_op-inl.h

```diff
@@ -1563,6 +1563,186 @@ inline void AdamUpdateEx(const nnvm::NodeAttrs& attrs,
+    float lower_bound;
+    float upper_bound;
+    float t;
+    bool bias_correction;
+    float wd;
+    float rescale_grad;
+    float clip_gradient;
+    DMLC_DECLARE_PARAMETER(LAMBParam) {
+      DMLC_DECLARE_FIELD(lr)
+      .describe("Learning rate");
+      DMLC_DECLARE_FIELD(beta1)
+      .set_default(0.9f)
+      .describe("The decay rate for the 1st moment estimates.");
+      DMLC_DECLARE_FIELD(beta2)
+      .set_default(0.999f)
+      .describe("The decay rate for the 2nd moment estimates.");
+      DMLC_DECLARE_FIELD(epsilon)
+      .set_default(1e-6f)
+      .describe("A small constant for numerical stability.");
+      DMLC_DECLARE_FIELD(lower_bound)
+      .set_default(1e-3f)
+      .describe("Lower limit of norm of weight.");
+      DMLC_DECLARE_FIELD(upper_bound)
+      .set_default(10.0f)
+      .describe("Upper limit of norm of weight.");
+      DMLC_DECLARE_FIELD(t)
+      .describe("Index update count.");
+      DMLC_DECLARE_FIELD(bias_correction)
+      .set_default(false)
+      .describe("Whether to use bias correction.");
+      DMLC_DECLARE_FIELD(wd)
+      .describe("Weight decay augments the objective function with a "
+                "regularization term that penalizes large weights. "
+                "The penalty scales with the square of the magnitude of each weight.");
+      DMLC_DECLARE_FIELD(rescale_grad)
+      .set_default(1.0f)
+      .describe("Rescale gradient to grad = rescale_grad*grad.");
+      DMLC_DECLARE_FIELD(clip_gradient)
+      .set_default(-1.0f)
+      .describe("Clip gradient to the range of [-clip_gradient, clip_gradient] "
+                "If clip_gradient <= 0, gradient clipping is turned off. "
+                "grad = max(min(grad, clip_gradient), -clip_gradient).");
+    }
+};
+
+struct LAMBWeightParam : public dmlc::Parameter<LAMBWeightParam> {
+    float lr;
+    float r1;
+    float r2;
+    float lower_bound;
+    float upper_bound;
+    bool bias_correction;
+    DMLC_DECLARE_PARAMETER(LAMBWeightParam) {
+      DMLC_DECLARE_FIELD(lr)
+      .describe("Learning rate");
+      DMLC_DECLARE_FIELD(r1)
+      .describe("r1");
+      DMLC_DECLARE_FIELD(r2)
+      .describe("r2");
+      DMLC_DECLARE_FIELD(lower_bound)
+      .set_default(1e-3f)
+      .describe("Lower limit of norm of weight.");
+      DMLC_DECLARE_FIELD(upper_bound)
+      .set_default(10.0f)
+      .describe("Upper limit of norm of weight.");
+      DMLC_DECLARE_FIELD(bias_correction)
+      .set_default(false)
+      .describe("Whether to use bias correction.");
+    }
+};
+
+struct LAMBUpdateKernel {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data,
+    DType* mean_data, DType* var_data, const DType* weight_data, const DType* grad_data,
+    const DType clip_gradient, const DType rescale_grad,
+    const DType beta1, const DType beta2,
+    DType lr, const DType wd,
+    const DType epsilon, const DType lower_bound,
+    const DType upper_bound, const DType t,
```

Comment:
are upper_bound, lower_bound, etc being used at all??? Remove them from the argument list if they are not required